chainstate/src/detail/mod.rs

```diff
@@ -744,28 +748,38 @@ impl<S: BlockchainStorage, V: TransactionVerificationStrategy> Chainstate<S, V>
         &self.events_controller
     }
 
-    pub fn is_initial_block_download(&self) -> Result<bool, PropertyQueryError> {
-        if self.is_initial_block_download_finished {
-            return Ok(false);
+    pub fn is_initial_block_download(&self) -> bool {
+        !self.is_initial_block_download_finished.test()
+    }
+
+    /// Returns true if the given block timestamp is newer than `ChainstateConfig::max_tip_age`.
+    fn is_fresh_block(&self, time: &BlockTimestamp) -> bool {
+        let now = self.time_getter.get_time();
+        time.as_duration_since_epoch() + self.chainstate_config.max_tip_age.clone().into() > now
+    }
+
+    /// Update `is_initial_block_download_finished` when tip has changes (can only be set once)
+    fn update_initial_block_download_flag(&mut self) -> Result<(), PropertyQueryError> {
+        if self.is_initial_block_download_finished.test() {
+            return Ok(());
         }
 
         // TODO: Add a check for importing and reindex.
 
         // TODO: Add a check for the chain trust.
 
         let tip_timestamp = match self.query()?.get_best_block_header() {
-            Ok(h) => Ok(h.timestamp()),
+            Ok(h) => h.timestamp(),
             // There is only the genesis block, so the initial block download isn't finished yet.
-            Err(PropertyQueryError::GenesisHeaderRequested) => return Ok(true),
-            Err(e) => Err(e),
-        }?;
-        Ok(!self.is_fresh_block(&tip_timestamp))
-    }
+            Err(PropertyQueryError::GenesisHeaderRequested) => return Ok(()),
+            Err(e) => return Err(e),
+        };
 
-    /// Returns true if the given block timestamp is newer than `ChainstateConfig::max_tip_age`.
-    fn is_fresh_block(&self, time: &BlockTimestamp) -> bool {
-        let now = self.time_getter.get_time();
-        time.as_duration_since_epoch() + self.chainstate_config.max_tip_age.clone().into() > now
+        if self.is_fresh_block(&tip_timestamp) {
+            self.is_initial_block_download_finished.set();
+        }
+
+        Ok(())
```

Comment:
I wonder whether we should write `is_initial_block_download_finished` to the DB once it becomes true, so that if the node is relaunched, it doesn't get reset to false again.

Reply:
It should be reset when the node is shutdown, because the node could be down for a while and hence the "redownload" process should be started again.

Reply:
> It should be reset when the node is shutdown, because the node could be down for a while and hence the "redownload" process should be started again.

Something similar can happen when the network is down for a while but the node keeps on running. Once we're reconnected, the node has to catch up. How should that be handled? Seems the current PR prevents the node from reverting into the IBD stage in that case but maybe it should be handled somehow too.

Reply:
Under normal operation circumstances, the node should never fall back. Besides, the IBD state should be seen as an optimization, more than it being a global state.